src/replication.c

```diff
@@ -500,6 +499,59 @@ trigger_rebuild(spec_t *spec)
 	}
 }
 
+static bool
+is_replica_familiar(spec_t *spec, replica_t *new_replica)
+{
+	known_replica_t *kr = NULL;
+	int familiar_replicas = 0;
+	bool ret = false, found = false;
+	replica_t *old_replica = NULL;
+
+	ASSERT(MTX_LOCKED(&spec->rq_mtx));
+
+	TAILQ_FOREACH(kr, &spec->identified_replica, next) {
+		familiar_replicas++;
+		if (kr->zvol_guid == new_replica->zvol_guid) {
+			found = true;
+			if (!kr->is_connected) {
+				ret = true;
+				kr->is_connected = true;
+			}
+			break;
+		}
+	}
+
+	if (!found && familiar_replicas < spec->replication_factor) {
+		kr = malloc(sizeof (known_replica_t));
+		kr->zvol_guid = new_replica->zvol_guid;
+		kr->is_connected = true;
+		ret = true;
+		TAILQ_INSERT_TAIL(&spec->identified_replica, kr, next);
+	}
+
+	/*
+	 * If the target has already an active session with this replica
+	 * then the target will disconnect from the previous session and
+	 * allow a new session
+	 */
+	if (found && !ret) {
+		TAILQ_FOREACH(old_replica, &spec->rq, r_next) {
+			if (old_replica->zvol_guid == new_replica->zvol_guid) {
+				MTX_UNLOCK(&spec->rq_mtx);
+				REPLICA_LOG("Disconnecting replica(%lu)'s old "
+				    "session\n", old_replica->zvol_guid);
+				handle_mgmt_conn_error(old_replica, 0, NULL, 0);
```

Comment:
during this, what happens for any pending events received in 'init_replication' thread?

Reply:
Good find!
We will not allow a new connection due to complexity. At this point, we won't be able to know any pending events on old replica session.